protostar/configuration_file/configuration_file_factory.py

```diff
@@ -0,0 +1,85 @@
+from pathlib import Path
+from typing import Optional
+
+from protostar.protostar_exception import ProtostarException
+
+from .configuration_file import ConfigurationFile
+from .configuration_file_v1 import ConfigurationFileV1
+from .configuration_file_v2 import ConfigurationFileV2
+from .configuration_legacy_toml_interpreter import ConfigurationLegacyTOMLInterpreter
+from .configuration_toml_interpreter import ConfigurationTOMLInterpreter
+
+
+class ConfigurationFileFactory:
+    def __init__(self, cwd: Path) -> None:
+        self._cwd = cwd
+
+    def create(self) -> Optional[ConfigurationFile]:
+        protostar_toml_path = self._search_upwards_protostar_toml_path()
+        if protostar_toml_path is None:
+            return None
+        protostar_toml_content = protostar_toml_path.read_text()
+
+        configuration_file_v2 = self._create_configuration_toml_v2(
+            protostar_toml_path,
+            protostar_toml_content,
+        )
+        if configuration_file_v2:
+            return configuration_file_v2
+
+        configuration_file_v1 = self._create_configuration_toml_v1(
+            protostar_toml_path, protostar_toml_content
+        )
+        if configuration_file_v1:
+            return configuration_file_v1
+
+        raise ProtostarException(
+            f"{protostar_toml_path} must specify `protostar-version` for example:\n"
+            """
+            [project]
+            protostar-version = ?.?.?
+            """,
+        )
+
+    def _create_configuration_toml_v2(
```

Comment:
Shouldn't we have `AbstractConfigFactory`, `V1ConfigFactory` and `V2ConfigFactory`, imo factory shouldn't be an entity responsible for making decisions.

Reply:
My point is I wouldn't name it factory if you do not have a interchangeable family of those

Reply:
https://www.oreilly.com/library/view/head-first-design/0596007124/ch04.html
Chapter "Building a simple pizza factory"

In that chapter, the factory decides what Pizza needs to be build based on the provided type. In this case, the factory decides which configuration file should be created.

> My point is I wouldn't name it factory if you do not have a interchangeable family of those

Factory and Abstract Factory are two different things. In the link above, they use the Factory suffix even if the Factory is not abstract.